### PR TITLE
UI-side support for request.host information

### PR DESCRIPTION
**Pull request by jshaughn**
Adds an additional Tab in the edge summary to show requested host information, by response status code. 

This video shows external requests being routed to a serviceEntry, PassthroughCluster, and Unknown (i.e. BlackholeCluster).  With the **Hosts Tab** the user can now see exactly which requested host resulted in these endpoints.  Although not shown here, this information is available on all edges, revealing if nothing else, the full host name for the resulting service (e.g. reviews:9080):

kiali#1405

Server PR: https://github.com/kiali/kiali/pull/1687
Website PR: https://github.com/kiali/kiali.io/pull/167

Closes https://github.com/kiali/kiali/issues/1405

### CODE REVIEW

**Comment by josejulio** [comment]
Just pointing out that it doesn't work on TCP edges. 
I suppose is expected as we don't have any host information at that point (i think).

Screenshot_2019-09-19_12-38-56

**Comment by josejulio** [comment]
Peek 2019-09-19 17-03

Minor, i see that `PassthroughCluster` displays no hosts and a reference to the FAQ. Is it expected ? (I already configured Istio)

File: src/pages/Graph/SummaryPanelEdge.tsx [hunk 1]
```diff
@@ -167,8 +171,11 @@ export default class SummaryPanelEdge extends React.Component<SummaryPanelPropTy
                       </>
                     )}
                   </TabPane>
-                  <TabPane eventKey="responses" mountOnEnter={true} unmountOnExit={true}>
-                    <ResponseTable title={isGrpc ? 'GRPC codes:' : 'HTTP codes:'} responses={edgeData.responses} />
+                  <TabPane eventKey="flags" mountOnEnter={true} unmountOnExit={true}>
+                    <ResponseFlagsTable title={isGrpc ? 'GRPC codes:' : 'HTTP codes:'} responses={edgeData.responses} />
+                  </TabPane>
+                  <TabPane eventKey="hosts" mountOnEnter={true} unmountOnExit={true}>
+                    <ResponseHostsTable title={isGrpc ? 'GRPC codes:' : 'HTTP codes:'} responses={edgeData.responses} />
```
Comment (josejulio) [inline src/pages/Graph/SummaryPanelEdge.tsx]:
```suggestion
                    <ResponseHostsTable title="Hosts" responses={edgeData.responses} />
```

File: src/types/Graph.ts [hunk 2]
```diff
@@ -90,12 +90,11 @@ export interface CytoscapeMouseOutEvent extends CytoscapeBaseEvent {}
...
-// { code0: {
-//     flags0: %traffic,
-//     flags1: %traffic
-//   }}
-export type Responses = object;
+export type ResponseDetail = {
+  flags: object; // string map flags->percentageOfTraffic
+  hosts: object; // string map host->percentageOfTraffic
+};
+export type Responses = object; // map responseCode:string -> ResponseDetail
```
Comment (josejulio) [inline src/types/Graph.ts]: I would make this block a bit more specific and remove the `object` with index signatures i.e.
```typescript
// Graph Structures

type PercentageOfTrafficByFlag = {
  [flag: string]: string;
};

type PercentageOfTrafficByHost = {
  [host: string]: string;
};

export type ResponseDetail = {
  flags: PercentageOfTrafficByFlag;
  hosts: PercentageOfTrafficByHost;
};

export type Responses = {
  [responseCode: string]: ResponseDetail;
};

```

File: src/components/SummaryPanel/ResponseHostsTable.tsx [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+                    <td>{row.code}</td>
+                    <td>{row.host}</td>
+                    <td>{row.val}</td>
+                  </tr>
+                ))}
+              </tbody>
+            </table>
+          </>
+        ) : (
+          <>No Host Information (see FAQ)</>
```
Comment (josejulio) [inline src/components/SummaryPanel/ResponseHostsTable.tsx]: Just a reminder of this placeholder :)

File: src/pages/Graph/SummaryPanelEdge.tsx (hunk 1 above)
Comment (jshaughn) [reply src/pages/Graph/SummaryPanelEdge.tsx]: I updated the titles.

File: src/types/Graph.ts (hunk 2 above)
Comment (jshaughn) [reply src/types/Graph.ts]: +1, done.

File: src/components/SummaryPanel/ResponseHostsTable.tsx (hunk 3 above)
Comment (jshaughn) [reply src/components/SummaryPanel/ResponseHostsTable.tsx]: The FAQ PR has been closed and this message has been updated to be "No Host Information".

**Comment by jshaughn** [comment]
> Just pointing out that it doesn't work on TCP edges.

@josejulio I have added TCP support.  Here is an updated video:

kiali#1405-2

**Comment by jshaughn** [comment]
> Minor, i see that PassthroughCluster displays no hosts and a reference to the FAQ. Is it expected 

The FAQ reference is gone but because the edge in your example above did not have any traffic there will not be any host information (or traffic, etc).  So it would now say "No host information available" or something like that.

**Pull request merged**